src/Microsoft.AspNet.Authentication.OAuth/OAuthAuthenticationHandler.cs

```diff
@@ -159,14 +160,44 @@ protected override AuthenticationTicket AuthenticateCore()
             return new TokenResponse(oauth2Token);
         }
 
-        protected virtual async Task<AuthenticationTicket> GetUserInformationAsync(AuthenticationProperties properties, TokenResponse tokens)
+        protected virtual async Task<AuthenticationTicket> CreateTicketAsync(AuthenticationProperties properties, TokenResponse tokens)
         {
-            var context = new OAuthGetUserInformationContext(Context, Options, Backchannel, tokens)
+            var identity = new ClaimsIdentity(Options.ClaimsIssuer);
+
+            if (Options.SaveTokens)
             {
-                Properties = properties,
+                identity.AddClaim(new Claim("access_token", tokens.AccessToken, ClaimValueTypes.String, Options.ClaimsIssuer));
+
+                if (!string.IsNullOrEmpty(tokens.RefreshToken))
+                {
+                    identity.AddClaim(new Claim("refresh_token", tokens.RefreshToken, ClaimValueTypes.String, Options.ClaimsIssuer));
+                }
+
+                if (!string.IsNullOrEmpty(tokens.TokenType))
+                {
+                    identity.AddClaim(new Claim("token_type", tokens.TokenType, ClaimValueTypes.String, Options.ClaimsIssuer));
+                }
+
+                if (!string.IsNullOrEmpty(tokens.ExpiresIn))
+                {
+                    identity.AddClaim(new Claim("expires_in", tokens.ExpiresIn, ClaimValueTypes.String, Options.ClaimsIssuer));
+                }
+            }
+
+            var notification = new OAuthCreateTicketContext(Context, Options, Backchannel, tokens)
+            {
+                Principal = new ClaimsPrincipal(identity),
+                Properties = properties
             };
-            await Options.Notifications.GetUserInformationAsync(context);
-            return new AuthenticationTicket(context.Principal, context.Properties, Options.AuthenticationScheme);
+            
+            await Options.Notifications.CreateTicket(notification);
```

Comment:
nit: typically public async methods need to end with Async (not my personal fav, but wearing my @divega hat today)